rasa/core/policies/flow_policy.py

```diff
@@ -594,6 +598,17 @@ def _run_step(
             action_name = self.render_template_variables(step.action, context)
             if action_name in self.domain.action_names_or_texts:
                 structlogger.debug("flow.step.run.action", context=context)
+
+                if action_name.startswith("utter_ask_"):
```

Comment:
Is it intentional that we only do it for utter templates? I think it might make sense to update the counter no matter what (we can still ignore its effect later if we want) - might be a helpful information for a custom implementation as well.

Reply:
Yep, this was intentional, as per received feedback here, I thought this only makes sense for `collect_information` steps.

> might be a helpful information for a custom implementation as well.

Not sure I understand the value for updating the counter for any bot action, and also what a custom implementation refers to here?

Reply:
I was just saying that the counter, how often did we try to collect this information, is helpful in general. Maybe we can do the `action_name.startswith` part in the pattern and branch there? not sure though

Reply:
Oh so instead of counting `utter_ask`, we'd be counting the attempts for that `collect_information` step? in other words, how many times can the underlying pattern_ask_collect_information flow can run? I'll see if this simplifies implementation or not.

Reply:
I've tried testing after the new fixes by removing this `if` check, however found out that the invalid path tests that I added fail when the hardcoded number in `pattern_ask_collect_information`remains 2. Because now the counter stands for the number of bot actions taken (internal and end user facing), rather than how many times did the bot ask for a slot, it becomes less clear what should the max number of retries. I suggest keeping this if check, especially if in the future we allow users to customize the max number of retries.

Reply:
I'd recommend to use the step's id rather than the checking for the utter_ask prefix. we have a feature request to allow a user to specify a different action to use which potentially could not start with utter_ask

Reply:
Is it this feature request: https://rasahq.atlassian.net/browse/ATO-1567?

Reply:
yes